### ISSUE TITLE
handshake and 0rtt data in a single packet

### ISSUE THREAD

**Issue by kazuho** [quicwg/base-drafts#630]
Current packet layout requires the handshake (which is conveyed in a Cleartext packet) and 0-RTT data to be sent as separate packets. In other words, you need to send at least two packets to establish a connection and send a request (e.g. HTTP request, DNS query).

To me this seems like a regression when compared to TFO+TLS 1.3, in which you can convey SYN, TLS handshake and 0-RTT data in a single packet.

I'd argue that this might have real-world impact. Having the need to send two packets (instead of one) is not only costly but also has a higher chance of suffering from packet loss (especially in a lossy network), since the request cannot be processed on the server unless both of the two packets reach the server.

Therefore I think that we should find a way to send both the handshake and 0-RTT data in a single packet.

There could be several ways to accomplish this.

One way that might be worth considering would be to do the following.
* unify Cleartext type and 0-RTT encrypted type into one type
* at the beginning of the payload of the unified type include the amount of cleartext data. Cleartext data of the specified amount should immediately follow that. Rest of the payload would be 0-RTT data.

Since we would only send data of stream 0 in cleartext and we need to give precedence to the stream, storing cleartext before 0-RTT data makes sense.

**Comment by ianswett** [comment]
Excellent suggestion. I was just thinking about thinking about this this weekend, particularly for cases like DNS over QUIC.

I can see three framing options:
1) Create a new packet type which is cleartext + 0RTT data
2) Add two bytes at the beginning of cleartext packets to indicate a cleartext length that precedes the 0RTT data.
3) Require that all frames after a Stream 0 frame be encrypted with 0RTT data.

Option 3 seems overly clever, and 1 adds one or two more packet types.  Option 2 burns 2 bytes on every unencrypted packet, which I think may be the best option, but I may be wrong.

Any other framing suggestions?

**Comment by mikkelfj** [comment]
> Any other framing suggestions?

Just encrypt cleartext with 0RTT key and don't mind double encryption.

**Comment by kazuho** [comment]
> Any other framing suggestions?

My preference is 2. But any of the three would work.

And I will like to point out that we'd need to do the same for the first packets that the server sends. We need a packet format that contains both the payloads of current Server Cleartext and 1-RTT Encrypted (key phase 0). Otherwise, a server cannot respond with just sending _one_ packet.

The fact that a server might send multiple Cleartext packets (when doing a full handshake) might be considered a point against option 2 (i.e. burn 2 octets in every packet that includes Cleartext to indicate where Cleartext ends).

**Comment by kazuho** [comment]
> Just encrypt cleartext with 0RTT key and don't mind double encryption.

I am not sure if this approach works. A server might not know the 0RTT key, in which case it needs to fall back to doing a full handshake. For the purpose, handshake messages need to be sent unencrypted.

**Comment by ekr** [comment]
As @kazuho says, it does not work. The identification for the 0-RTT key is in the unencrypted portion, namely the ClientHello

**Comment by mikkelfj** [comment]
So encryption won't (always) work without extra retry fallback and roundtrip. I'd say option 2 sounds reasonable. The packet might be padded anyway to hide length of a HTTP request.

**Comment by ekr** [comment]
@kazuho: I agree it's annoying to have to send two packets.

I have a fourth option: Add a length field to the long header and allow multiple long headers to be packed into a single UDP datagram. Note that this will require adjusting the padding rules, but that should be possible.

**Comment by ianswett** [comment]
@ekr, conceptually I like option 4, but it introduces a bunch of weird options, such as the ability to specify two different connection IDs and packet numbers in the same packet.

If we did include 0RTT data in the same packet with a CHLO, we could make the plaintext the associated data, which would give it some extra protection from tampering, but we may want a hash as well, in case the 0RTT data can't be decrypted?

**Comment by ekr** [comment]
On Wed, Jun 14, 2017 at 4:14 PM, ianswett <notifications@github.com> wrote:

> @ekr <https://github.com/ekr>, conceptually I like option 4, but it
> introduces a bunch of weird options, such as the ability to specify two
> different connection IDs and packet numbers in the same packet.
>

What of it? Just handle the two packets as if they came in separate
datagrams.


> If we did include 0RTT data in the same packet with a CHLO, we could make
> the plaintext the associated data, which would give it some extra
> protection from tampering, but we may want a hash as well, in case the 0RTT
> data can't be decrypted?
>

This is starting to sound like a loot of hoop-jumping to avoid two calls to
send()

**Comment by janaiyengar** [comment]
I think this is a good idea in general. This will be particularly useful
for small requests, such as for DNS.

I think we can pull this off without any framing changes. The current
packet types (post-Paris) have the Client Initial packet always carrying
the initial handshake data. Since this packet will contain only a STREAM
frame with handshake material, we can make this packet a bit more special.
Since STREAM frames have a size field, any bytes beyond the cleartext
STREAM frame can be treated as 0-RTT encrypted data.

**Comment by janaiyengar** [comment]
I think this is a good idea in general. This will be particularly useful
for small requests, such as for DNS.

I think we can pull this off without any framing changes. The current
packet types (post-Paris) have the Client Initial packet always carrying
the initial handshake data. Since this packet will contain only a STREAM
frame with handshake material, we can make this packet a bit more special.
Since STREAM frames have a size field, any bytes beyond the cleartext
STREAM frame can be treated as 0-RTT encrypted data.

Thoughts?

**Comment by ekr** [comment]
On Wed, Jun 14, 2017 at 8:23 PM, janaiyengar <notifications@github.com>
wrote:

> I think this is a good idea in general. This will be particularly useful
> for small requests, such as for DNS.
>
> I think we can pull this off without any framing changes. The current
> packet types (post-Paris) have the Client Initial packet always carrying
> the initial handshake data. Since this packet will contain only a STREAM
> frame with handshake material, we can make this packet a bit more special.
> Since STREAM frames have a size field, any bytes beyond the cleartext
> STREAM frame can be treated as 0-RTT encrypted data.


And what if you need to pad to get up to 1280?

**Comment by janaiyengar** [comment]
Encrypt the padding. 0-RTT encrypted data can contain STREAM frame followed
by PADDING frames.

**Comment by lucas-clemente** [comment]
I think it's a nice property of QUIC that right now packetization (header & encryption) is separate from framing, which simplifies layering in implementations. If we do this, I think we should keep it on the public header level, so that it conceptually works like two packets in one.

**Comment by janaiyengar** [comment]
Yeah, that's a good point. So, then I'm leaning towards ekr's 4th option,
but not for all long header packets but only Client Initial packets. We can
add a 2-byte length field inside the payload of the Client Initial packet.

**Comment by kazuho** [comment]
> @ekr, conceptually I like option 4, but it introduces a bunch of weird options, such as the ability to specify two different connection IDs and packet numbers in the same packet.

I share the same concern as @ianswett.

If we go for option 4, the server would be required to check that the connection IDs included in the single packet, or that the types of succeeding packet headers are not Client Cleartext. Otherwise, it could be used as an attack vector to cause the server send multiple ServerHellos (of different conenction IDs) to a single machine.

So if we are to go for option 4 I would prefer omitting connection ID and type (which can be implied to be 0-RTT Encrypted) for the succeeding packet headers.

**Comment by ekr** [comment]
Why wouldn't it be enough to simply not allow two ClientInitials in the same packet?

**Comment by kazuho** [comment]
> Why wouldn't it be enough to simply not allow two ClientInitials in the same packet?

Unless there is a guarantee that a single packet does not contain information about more than one connection (i.e. either only one connection ID is included in a packet or all the connection IDs included in a packet are identical), it would be very difficult for a load balancer to route the packet to the correct backend.

That is why I believe that we need to have a requirement that all the connection IDs are identical if we choose option 4.

**Comment by mcmanus** [comment]
I think it makes sense to worry about smuggling-attack-bugs here... so the clear semantics of Ian's option 1 (new type) is the most attractive to me.

**Comment by mikkelfj** [comment]
> I think it makes sense to worry about smuggling-attack-bugs here... so the clear semantics of Ian's option 1 (new type) is the most attractive to me.

It also has the benefit of simplifing routing and optimizatoins for 0RTT logic to dedicated software / thread or even dedicated hosts while at the same time allowing an implementation to completely ignore 0RTT or 1RTT if so inclined. Also, it falls in line with the idea that the packet type is sufficient to know what keys are involved in packet.

**Comment by ekr** [comment]
proposal #1 actually seems extremely different from all other packet types. For instance, does it have an FNV1a authentication tag? If so, does that go *after* the 0-RTT ciphertext?

**Comment by mcmanus** [comment]
On Thu, Jun 15, 2017 at 1:27 PM, ekr <notifications@github.com> wrote:

> proposal #1 <https://github.com/quicwg/base-drafts/pull/1> actually seems
> extremely different from all other packet types.
>

indeed. but having 2 quic packets bundled under one recv() (via 2 long
headers) seems even more different to me - losing the transport 1:1
message-ness is surprising.

For instance, does it have an FNV1a authentication tag?
>

seems so


> If so, does that go *after* the 0-RTT ciphertext?
>

where is it critical (as long as its defined)?

**Comment by ekr** [comment]
OK, I guess I don't find it that surprising. That's a feature of DTLS, so it feels familiar to me.

And if we decide to AEAD-encrypt everything under a null key, we'll double-AEAD. ugh...

**Comment by siyengar** [comment]
I brought this up with @janaiyengar and @Ianswett at Chicago as well. I'm in favor of option 4 as well which is something I also suggested :)

**Comment by siyengar** [comment]
Without option 4 that is with a different packet type, we have to worry about what attacks are possible if a mitm being is able to change the initial packet type and then change the packet split. I don't think This has  implications to crypto but allowing the split could mess with transport parameter calculations

**Comment by MikeBishop** [comment]
Is there a reason not to simply define Client Initial to *always* contain a cleartext portion and a 0-RTT portion?  If you don't have a 0-RTT key, that section will be empty, but so what?  If the server doesn't accept the 0-RTT that you didn't offer, it won't try to decrypt it, so it can be random padding.  (Or you can pad in cleartext, I suppose.)

**Comment by ekr** [comment]
As noted above, this creates a situation in which there are two different packet protection mechanisms applied to the same message (the AEAD and the cleartext integrity check). That seems unfortunate.

**Comment by kazuho** [comment]
> As noted above, this creates a situation in which there are two different packet protection mechanisms applied to the same message (the AEAD and the cleartext integrity check). That seems unfortunate.

If that is an issue, we could define the range protected by FNV-1a to _not_ cover 0-RTT data, and the range protected by AAD to _not_ cover cleartext payload. Something like below. The ranges covered by FNV-1a and AAD are shown in the vertical bars in right. It also adds a "Cleartext Length" field.

```
+-+-+-+-+-+-+-+-+                                                    -    -
|1|   Type (7)  |                                                    |    |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    |    |
|                                                               |    |    |A
+                       Connection ID (64)                      +    |    |E
|                                                               |    |F   |A
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    |N   |D
|                       Packet Number (32)                      |    |V   |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    |-   |A
|                         Version (32)                          |    |1   |A
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    |a   |D
|    Cleartext Length (16)      |                                    |    |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    |    -
|           Cleartext Payload (Cleartext Length)              ...    |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+    -    
|                      Encrypted Payload (*)                  ...
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

**Comment by ianswett** [comment]
Coming back around on this, I think we end up needing something like option 4 for a simple reason.  We need the ability to ack the cleartext data and not ack the 0RTT data if 0RTT fails.  So they need different packet numbers, even if they're in the same UDP packet.  At that point, sticking together two QUIC 'packets' in a single UDP packet would seem the obvious approach.

**Comment by mikkelfj** [comment]
I could see multiple long headers in one Packet, but allowing for multiple packets in one MTU affects a lot of machinery.

**Comment by ianswett** [comment]
To build on that idea, avoid adding anything new(ie: a length) to the unencrypted public header, and make these packets look as much as possible like other QUIC packets, I propose we:
  * Add 4 frame types which indicate the remaining content is packet_number+1 for the same connection and the encryption level.  The header would be omitted.

In many cases this would mean encryption inside encryption, but I don't think that's a big problem in practice.

Admittedly, getting a framer to frame one of these is a bit of hassle, but it seems potentially useful for the HTTP case, since you may be able to send a request and/or header compression info along with the CHLO.  The CHLO is padded and head of line blocking for the entire connection, so this would be a strict performance win.

**Comment by martinthomson** [comment]
Encryption in encryption is a non-starter (for all the reasons we've discussed relating to the analysis of TLS, key usage, etc...).

**Comment by ianswett** [comment]
To clarify, encrypting something with 0RTT keys inside the handshake AES-GCM is not safe?  Any background about why, or is this a case of what's easy to prove?

**Comment by martinthomson** [comment]
The analysis of TLS (some analysis, not all), rely on strict separation of proofs.  That means that the handshake is assessed independent of the record layer.  One condition of that proof is that the keys that the handshake produces are never used to protect the handshake.  Otherwise, the proofs get entangled as a result of needing to show that the record layer doesn't abuse the keys in ways that would compromise the record layer.

Actually, maybe this is OK in the sense that the AES-GCM isn't really encryption, or related to the handshake at all.

But the point remains, it's easier to add a length to the long header and pack multiple records in the same datagram (oops, those are DTLS terms, see DTLS for an example of how this works) than to do double encryption.

**Comment by ianswett** [comment]
If we restrict this to embedding packets within the handshake packets encryption, this must be safe.  Because any observer could just apply the same encryption themselves.

And I'd be happy restricting it to handshake packets.  I'd even be happy just adding one frame type for 0RTT within handshake packets, unless someone comes up with a compelling use case for the others.  

I do think this is a better approach than adding a length, because adding a length introduces the possibility of bugs(ie: read length bytes into memory, attempt to 'de-frame' out of bounds memory, badness ensues) and it makes the 'header' larger and slightly more complex.

**Comment by kazuho** [comment]
Rethinking this, I now think that [option 4](https://github.com/quicwg/base-drafts/issues/630#issuecomment-308361033) (i.e. add length field to long packet header to allow multiple QUIC packets to be stored in a single UDP packet) is the simplest.

A server might reject 0-RTT data. In such case, INITIAL packet is accepted but 0-RTT data is rejected.

If we allow INITIAL packet to convey 0-RTT data, that would mean that the part of the packet can be acknowledged. That collides with the principle that sent data is acked per packet.

Option 4 avoids the issue by allowing two QUIC packets to fit into a single UDP packet.

**Comment by ianswett** [comment]
My suggestion was for the INITIAL packet to embed the 0RTT data in an encapsulated 'packet' with it's own packet number, though the embedded packet may have it's packet number defined by the fact it's embedded.  The reason to embed rather than adding a length field is because it's very easy to define a frame type which indicates an embedded packet, and adding a length seems more complex and is typically not useful.

Also, the belief is that if you can't process the handshake data, there's no reason to try to process the embedded 0RTT data.

**Comment by kazuho** [comment]
Thank you for pointing out that. I agree with your observations.

OTOH, it might still be the case that option 4 is easier to implement, since you do not need to encrypt INITIAL packet after encrypting 0RTT data as an encapsulated packet.

**Comment by RyanAtGoogle** [comment]
@ianswett That makes the most sense to me as well.

**Comment by janaiyengar** [comment]
Discussed at editors' meeting. We seemed to agree on something like option 4, with some specific rules: create "compound packet types" which have a length, and encode HS+0RTT and HS+1RTT long header packet types. To keep packet processing simple, the long header can be repeated so that a receiver can process parts of the packet under different keys in a straightforward manner.

This doesn't have to be the final resolution, but we should drive this issue to an end.

@kazuho : Would you be interested in driving this? Would you be able to write up a PR with a proposed design (use the one above if you like that design) and propose to the wg?

**Comment by marten-seemann** [comment]
Do we need to change the invariants for that?
It seems like we could achieve this by adding a length field to the Long Header.

**Comment by MikeBishop** [comment]
No, this doesn't modify the invariants.  We'd just define additional packet types for v1 which have a length + additional packet to follow.

**Comment by kazuho** [comment]
@janaiyengar 
> Discussed at editors' meeting. We seemed to agree on something like option 4, with some specific rules: create "compound packet types" which have a length, and encode HS+0RTT and HS+1RTT long header packet types.
> (snip)
> @kazuho : Would you be interested in driving this? Would you be able to write up a PR with a proposed design (use the one above if you like that design) and propose to the wg?

I can handle that, but should we create a PR for this issue now?

I think that the design decision might change based on if we will be having separate packet number space for handshake and protected packets.

**Comment by ianswett** [comment]
I'm happy to see a PR.

I'm happy adding a length to the INITIAL and HANDSHAKE packet types and then another packet type could be put in the same packet.  I'd rather not repeat connection ID(s) because I don't believe it's possible for them to be different, but given I think we're heading for different PN spaces for unencrypted and encrypted packets, repeating the PN portion is important.

The negative of this is that the second packet type is a special case because you have to know it doesn't include connection IDs.  But if you include them, it's a waste of space and redundant.  I think I prefer a special case over redundant information that may not match, but that's me.

**Comment by mikkelfj** [comment]
Wasting space is hardly a concern in packets of 1280 bytes in length that are likely padded? And for long certs, if you need multiple packets, CID is hardly going to be the deciding factor. So I'd say keep the packet formats unchanged.

I would assume the packet number increases in the second packet added to the same MTU, no?

I think it is a good idea to only allow certain packets to be heading other packes such as Initial and Handshake, otherwise general packet handling gets unnecessarily complex.

**Comment by kazuho** [comment]
> Wasting space is hardly a concern in packets of 1280 bytes in length that are likely padded?

One common case where the space becomes precious is when you are doing 0 RTT, and the server tries to use a compound packet to send ServerHello...ServerFinished and the response in 0.5 RTT.

**Comment by martinthomson** [comment]
Fixed by #1262 \o/.  Thanks Kazuho.
